File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;
class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int size = friends.length;
        HashMap<String, Integer> hm = new HashMap<>();
        for(int i=0; i<size; i++) hm.put(friends[i],i);
        int map[][] = new int[size][size];
        for(String gift : gifts){
            String strs[] = gift.split(" ");
            int a = hm.get(strs[0]);
            int b = hm.get(strs[1]);
            map[a][b]++;
        }
        int score[] = new int[size];
        for(int i=0; i<size; i++){
            for(int j=0; j<size; j++){
                score[i] += map[i][j];
                score[i] -= map[j][i];
            }
        }
        int receive[] = new int[size];
        for(int i=0; i<size; i++){
            for(int j=0; j<size; j++){
                if(i!=j && map[i][j] > map[j][i]) receive[i]++;
                else if(i!=j && map[i][j] == map[j][i]){
                    if(score[i] > score[j]) receive[i]++;
                }
            }
        }
        for(int v : receive) answer = Math.max(answer,v);
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cpp
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;
unordered_map<string, int> 선물지수;
unordered_map<string, int> nameToNum;
vector<vector<int>> 선물기록;
bool compare(string f1, string f2){
    int score1 = 선물기록[nameToNum[f1]][nameToNum[f2]];
    int score2 = 선물기록[nameToNum[f2]][nameToNum[f1]];
    if(score1 == score2){
        if(선물지수[f1] > 선물지수[f2]) return true;
    }else if(score1 > score2) return true;
    return false;
}

int solution(vector<string> friends, vector<string> gifts) {
    for(int i=0; i<friends.size(); i++){
        nameToNum[friends[i]] = i;
    }
    선물기록.resize(friends.size(), vector<int>(friends.size(),0));
    for(auto gift : gifts){
        int index = gift.find(" ");
        string from = gift.substr(0, index);
        string to = gift.substr(index+1);
        선물지수[from]++;
        선물지수[to]--;
        선물기록[nameToNum[from]][nameToNum[to]]++;
    }
    int maxValue = 0;
    for(int i=0; i<friends.size(); i++){
        int ans = 0;
        for(int j=0; j<friends.size(); j++){
            if(i==j) continue;
            if(compare(friends[i], friends[j])) ans++;
        }
        maxValue = max(maxValue, ans);
    }
    return maxValue;
}